File: sushigraphics/sushigraphics/GraphicsDevice.swift
//
//  GraphicsDevice.swift
//  sushigraphics
//
//  Created by Michael Ong on 28/6/17.
//  Copyright © 2017 Michael Ong. All rights reserved.
//

import Foundation

import sushicore

public class GraphicsDevice {

    public var      driver          : GraphicsDeviceDriver?


    public internal(set)
    var             backingDevice   : Any


    public func     createSurface   (with initialSize: Vector2, _ dpi: Double, to target: Any) throws -> GraphicsSurface {

        guard let driver = driver else { throw FastFailType.driverNotInstalled }

        let     surface                 = GraphicsSurface()

                surface.backingSurface  = driver.createGraphicsSurface(with: initialSize, dpi, to: target)
                surface.driver          = driver

                surface.backingSize     = initialSize

        return  surface
    }


    public init                     (backingType: Any) {

        self.backingDevice = backingType
    }
}

File: sushicpp/sushigraphics/sushigraphics/impl/graphicsDevice.cpp
//
//  graphicsDevice.cpp
//  sushigraphicsios
//
//  Created by Michael Ong on 19/8/17.
//  Copyright © 2017 Michael Ong. All rights reserved.
//

#include "../graphicsDevice.hpp"
#include "../graphicsSurface.hpp"

#include <sushicore/FastFail.h>

using namespace sushi::core;
using namespace sushi::graphics;

using namespace sushi::drivers;

GraphicsDevice::GraphicsDevice                      (GraphicsDriver* driver):

CoreObject      (),

driver          (driver),
backingDevice   (driver->createGraphicsDevice()) {

}

GraphicsDevice::~GraphicsDevice                     () {

    backingDevice   = NULL;
    driver          = nullptr;
}

GraphicsSurface*    GraphicsDevice::createSurface   (const glm::vec2 &initialSize, const float &initialDpi, void *backingWindow) const {

    if (driver == nullptr) {

        FastFail::instance->crash(failureTypes::driverNotInstalled("Graphics Device driver not installed!"));
    }

    auto    surface                 = new GraphicsSurface();

            surface->graphicsDriver = driver;

            surface->backingSurface = driver->createBackingSurface(initialSize, initialDpi, backingWindow);
            surface->backingSize    = initialSize;

    return  surface;
}
